File: WordPress/src/main/java/org/wordpress/android/WordPress.java
package org.wordpress.android;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.NonNull;

import com.android.volley.RequestQueue;

import org.wordpress.android.fluxc.tools.FluxCImageLoader;
import org.wordpress.android.modules.AppComponent;
import org.wordpress.android.modules.DaggerAppComponent;
import org.wordpress.android.networking.RestClientUtils;
import org.wordpress.android.util.AppLog;
import org.wordpress.android.util.BitmapLruCache;

import java.lang.reflect.Field;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.HasAndroidInjector;

public class WordPress extends BaseWordPress implements HasAndroidInjector {
    public static final String SITE = "SITE";
    public static final String LOCAL_SITE_ID = "LOCAL_SITE_ID";
    public static final String REMOTE_SITE_ID = "REMOTE_SITE_ID";
    public static String versionName;
    public static WordPressDB wpDB;
    public static boolean sAppIsInTheBackground = true;

    @Inject DispatchingAndroidInjector<Object> mDispatchingAndroidInjector;

    @Inject AppInitializer mAppInitializer;

    public static RequestQueue sRequestQueue;
    public static FluxCImageLoader sImageLoader;

    @Override public AndroidInjector<Object> androidInjector() {
        return mDispatchingAndroidInjector;
    }

    @NonNull @Override public AppInitializer initializer() {
        return mAppInitializer;
    }

    public AppComponent component() {
        return appComponent;
    }

    public static BitmapLruCache getBitmapCache() {
        return AppInitializer.Companion.getBitmapCache();
    }

    public static Context getContext() {
        return AppInitializer.Companion.getContext();
    }

    protected void initDaggerComponent() {
        appComponent = DaggerAppComponent.builder()
                                         .application(this)
                                         .build();
    }

    public static void updateContextLocale() {
        AppInitializer.Companion.updateContextLocale();
    }

    public static RestClientUtils getRestClientUtils() {
        return AppInitializer.Companion.getRestClientUtils();
    }

    public static RestClientUtils getRestClientUtilsV1_1() {
        return AppInitializer.Companion.getRestClientUtilsV1_1();
    }

    public static RestClientUtils getRestClientUtilsV1_2() {
        return AppInitializer.Companion.getRestClientUtilsV1_2();
    }

    public static RestClientUtils getRestClientUtilsV1_3() {
        return AppInitializer.Companion.getRestClientUtilsV1_3();
    }

    public static RestClientUtils getRestClientUtilsV2() {
        return AppInitializer.Companion.getRestClientUtilsV2();
    }

    public static RestClientUtils getRestClientUtilsV0() {
        return AppInitializer.Companion.getRestClientUtilsV0();
    }

    public static String getDefaultUserAgent() {
        return AppInitializer.Companion.getDefaultUserAgent();
    }

    public static String getUserAgent() {
        return AppInitializer.Companion.getUserAgent();
    }

    @Override
    public void onCreate() {
        super.onCreate();

        // Init Dagger
        initDaggerComponent();
        component().inject(this);

        mAppInitializer.init();
    }

    public static final String USER_AGENT_APPNAME = "wp-android";

    /**
     * Gets a field from the project's BuildConfig using reflection. This is useful when flavors
     * are used at the project level to set custom fields.
     * based on: https://code.google.com/p/android/issues/detail?id=52962#c38
     *
     * @param application Used to find the correct file
     * @param fieldName The name of the field-to-access
     * @return The value of the field, or {@code null} if the field is not found.
     */
    public static Object getBuildConfigValue(Application application, String fieldName) {
        try {
            String packageName = application.getClass().getPackage().getName();
            Class<?> clazz = Class.forName(packageName + ".BuildConfig");
            Field field = clazz.getField(fieldName);
            return field.get(null);
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Gets a field from the project's BuildConfig using reflection. This is useful when flavors
     * are used at the project level to set custom fields.
     * based on: https://code.google.com/p/android/issues/detail?id=52962#c38
     *
     * @param activity Used to get the Application instance
     * @param configValueName The name of the field-to-access
     * @return The string value of the field, or empty string if the field is not found.
     */
    public static String getBuildConfigString(Activity activity, String configValueName) {
        if (!BuildConfig.DEBUG) {
            return "";
        }

        String value = (String) WordPress.getBuildConfigValue(activity.getApplication(), configValueName);
        if (!TextUtils.isEmpty(value)) {
            AppLog.d(AppLog.T.NUX, "Auto-filled from build config: " + configValueName);
            return value;
        }

        return "";
    }
}

File: WordPress/src/main/java/org/wordpress/android/WordPress.kt
package org.wordpress.android

import android.app.Application
import android.content.Context
import android.database.SQLException
import android.os.Build
import android.util.Log
import androidx.work.Configuration
import coil.decode.VideoFrameDecoder
import com.android.volley.RequestQueue
import dagger.hilt.EntryPoints
import okhttp3.OkHttpClient
import org.wordpress.android.fluxc.tools.FluxCImageLoader
import org.wordpress.android.modules.AppComponent
import org.wordpress.android.ui.main.EdgeToEdgeActivityLifecycleCallbacks
import org.wordpress.android.util.AppLog
import org.wordpress.android.util.AppLog.T

/**
 * An abstract class to be extended by {@link WordPressApp} for real application and WordPressTest for UI test
 * application. Containing public static variables and methods to be accessed by other classes.
 */
abstract class WordPress : Application(), coil.ImageLoaderFactory, Configuration.Provider {
    abstract fun initializer(): AppInitializer

    /**
     * WorkManager is initialized on demand (the default initializer is removed in the manifest) so that its
     * Room database is opened by whichever thread first needs it, off the main thread on normal process starts.
     */
    override val workManagerConfiguration: Configuration
        get() = Configuration.Builder()
            .setWorkerFactory(component().wordPressWorkersFactory())
            .setJobSchedulerJobIdRange(WORK_MANAGER_ID_RANGE_MIN, WORK_MANAGER_ID_RANGE_MAX)
            .apply { if (BuildConfig.DEBUG) setMinimumLoggingLevel(Log.DEBUG) }
            .build()

    override fun onCreate() {
        super.onCreate()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.VANILLA_ICE_CREAM) {
            registerActivityLifecycleCallbacks(
                EdgeToEdgeActivityLifecycleCallbacks()
            )
        }
    }

    fun component(): AppComponent = EntryPoints.get(this, AppComponent::class.java)

    fun wordPressComSignOut() {
        initializer().wordPressComSignOut()
    }

    /**
     * This returns a singleton Coil ImageLoader that's accessed with context.imageLoader
     */
    override fun newImageLoader(): coil.ImageLoader {
        val authUtils = component().authenticationUtils()
        val okHttpClient = OkHttpClient.Builder()
            .addInterceptor { chain ->
                val request = chain.request()
                val authHeaders = authUtils.getAuthHeaders(request.url.toString())
                val newRequest = request.newBuilder().apply {
                    authHeaders.forEach { (name, value) -> header(name, value) }
                }.build()
                chain.proceed(newRequest)
            }
            .build()
        return coil.ImageLoader.Builder(this)
            .okHttpClient(okHttpClient)
            .crossfade(true)
            .components {
                add(VideoFrameDecoder.Factory())
            }
            .build()
    }

    @Suppress("TooManyFunctions")
    companion object {
        const val SITE = "SITE"
        const val LOCAL_SITE_ID = "LOCAL_SITE_ID"
        const val REMOTE_SITE_ID = "REMOTE_SITE_ID"
        const val USER_AGENT_APPNAME = "wp-android"

        // Use service ids near the int max to avoid collisions with existing JobService ids
        // The minimum range size is 1000, but we can easily give 10000.
        private const val WORK_MANAGER_ID_RANGE_MAX = Int.MAX_VALUE
        private const val WORK_MANAGER_ID_RANGE_MIN = WORK_MANAGER_ID_RANGE_MAX - 10000

        lateinit var versionName: String
        var appIsInTheBackground = true

        /**
         * The legacy app database. Opened lazily (and warmed up off the main thread by [AppInitializer]) so
         * that background process starts don't run the SQLite open + migrations on the main thread.
         */
        @JvmStatic
        val wpDB: WordPressDB by lazy { openWpDb(getContext()) }

        private fun openWpDb(context: Context): WordPressDB {
            return try {
                WordPressDB(context)
            } catch (e: SQLException) {
                AppLog.e(T.DB, "Invalid database, deleting and recreating it", e)
                WordPressDB.deleteDatabase(context)
                WordPressDB(context)
            }
        }

        @JvmField
        var requestQueue: RequestQueue? = null

        @JvmField
        var imageLoader: FluxCImageLoader? = null

        @JvmStatic
        fun getBitmapCache() = AppInitializer.getBitmapCache()

        @JvmStatic
        fun getContext() = AppInitializer.context!!

        @JvmStatic
        fun getRestClientUtils() = AppInitializer.restClientUtils

        @Suppress("FunctionNaming")
        @JvmStatic
        fun getRestClientUtilsV1_1() = AppInitializer.restClientUtilsV1_1

        @Suppress("FunctionNaming")
        @JvmStatic
        fun getRestClientUtilsV1_2() = AppInitializer.restClientUtilsV1_2

        @Suppress("FunctionNaming")
        @JvmStatic
        fun getRestClientUtilsV1_3() = AppInitializer.restClientUtilsV1_3

        fun getRestClientUtilsV2() = AppInitializer.restClientUtilsV2

        fun getRestClientUtilsV0() = AppInitializer.restClientUtilsV0
    }
}
